### PR TITLE
Fix emptying cart issue for custom order statuses

### PR DESCRIPTION
**Pull request by alexedwardjones**
### Changes proposed in this Pull Request:

I was experiencing an issue where a custom status wasn't being picked up and causing the cart to be emptied.

After some discussion, I've changed the cart emptying to be based on a boolean value and added a filter for the boolean. The filter takes two parameters, the default boolean value based on the order status and the order in case third-parties want to do further checks.

#### Recreating the issue
1. Install [the Stripe plugin](https://wordpress.org/plugins/woocommerce-gateway-stripe/)
2. Activate AliPay via Stripe. It doesn't need to configured we just need a payment method that doesn't complete immediately, in this case an external payment gateway.
1. Install [the Code Snippets plugin](https://wordpress.org/plugins/code-snippets/)
3. Add and activate the below code snippet which adds a custom order status `Foo` and makes it the default status.
```
function register_foo_post_status() {
	register_post_status(
		'wc-foo',
		array(
			'label'   => 'Foo',
			'public'	=> true,
			'show_in_admin_status_list' => true,
			'label_count'               => _n_noop( 'Foo <span class="count">(%s)</span>', 'Foo <span class="count">(%s)</span>' )
		)
	);

}
add_action( 'init', 'register_foo_post_status' );

function register_foo_order_status( $statuses ) {
	$statuses['wc-foo'] = 'Foo';

	return $statuses;
}
add_filter( 'wc_order_statuses', 'register_foo_order_status' );

function register_foo_order_post_status( $statuses ) {
	$statuses['wc-foo'] = 'Foo';
	return $statuses;
}
add_filter( 'woocommerce_register_shop_order_post_statuses', 'register_foo_order_post_status' );

function new_default_order_status( $status ) {
	return 'foo';
}
add_filter( 'woocommerce_default_order_status', 'new_default_order_status' );

function add_foo_as_payment_status( $statuses ) {
	$statuses[] = 'foo';
	return $statuses;
}
add_filter( 'woocommerce_valid_order_statuses_for_payment', 'add_foo_as_payment_status' );
```
4. Add a product to your cart.
5. Go through checkout with the "Alipay" payment method. The redirect to Alipay should fail because it's not configured but an order will be created in the new default status.
6. Navigating anywhere else on the site will **incorrectly** clear your cart because the new order status isn't picked up as being for an unpaid order.

#### Resolving the issue using the new filter
7. Add and activate a code snippet with the following code:
```
function dont_clear_cart_for_orders_with_foo_status( $did_order_succeed, $order ) {
  if ( 'foo' === $order->get_status() ) {
    return false;
  }

  return $did_order_succeed;
}
add_filter( 'woocommerce_should_clear_cart_after_payment',  'dont_clear_cart_for_orders_with_foo_status', 10, 2 );
```
8. Change the products in your cart, so that your cart total will be different and a new order will be generated when you attempt to pay.
9. Go through checkout with the "Alipay" payment method. The redirect to Alipay should fail because it's not configured but an order will be created in the new default status.
10. Navigating anywhere else on the site will no longer clear your cart because the new order status is now picked up as being for an unpaid order.

### CODE REVIEW

**Review by lsinger — CHANGES_REQUESTED** [review]
@alexedwardjones thanks for your contribution! Before diving more into it, let me just clarify how a new filter would need to be documented. The lines directly before the call to `apply_filters` should have a (correctly indented) comment block similar to this one: 

```
/*
 * A sentence describing what the filter does and why. In what situation would you want to use it?
 *
 * Optionally another sentence that discusses an edge case or some special use case.
 *
 * @since x.x.x (replace with WooCommerce version this will be included in)
 *
 * @param array $statuses Description of the $statuses parameter (one line like this for every parameter).
 *
 * @return array The filtered list of statuses.
 */
```

It also can't hurt to search the codebase for instances of `apply_filters` -- newer filters will usually have such a comment, while older filters may not have it yet.

Please re-request a review from me once you have added such a comment. Thanks!

**Comment by alexedwardjones** [comment]
Hey @lsinger 👋 Thanks, that was really helpful.

I've gone ahead and added that comment block. Let me know if it looks okay.

I'm not sure how your releases work so I wasn't sure what to put for the `@since` value. I saw you already have a release candidate for `8.6.0` so I put in `8.6.1`.

**Comment by alexedwardjones** [comment]
Hey @lsinger 👋 How do I go about getting this merged?

**Comment by alexedwardjones** [comment]
@lsinger @opr @jorgeatorres @Konamiman Hey folks 👋 Is anyone able to help me get this reviewed and merged?

**Review by lsinger — CHANGES_REQUESTED** [review]
Hello @alexedwardjones,

Sorry for taking some time here. We discussed this internally a bit and have come to the conclusion that a filter that changes the list of order statuses is a setup we're not comfortable with. 

If we support custom order statuses here one could reasonably expect for those to be supported elsewhere, too. However, the use case here is very focused, and at least for now -- until there's a clear road to a more holistic solution -- taking into account custom order statuses seems too far-reaching. 

Instead, we'd prefer a boolean filter in this case -- say, `woocommerce_should_clear_cart_after_payment` -- that would be able to take in the order and return whether the cart should be emptied.

**Comment by alexedwardjones** [comment]
Hey @lsinger 👋  Thanks for getting back to me. I've gone ahead and made the changes you're proposing. Let me know if it aligns with what you expected.

File: plugins/woocommerce/includes/wc-cart-functions.php
Comment (lsinger) [inline plugins/woocommerce/includes/wc-cart-functions.php]: We now have a double negation here, which makes it harder to reason about what this code does, so I'd like to remove that. Below is a suggestion as to how to go about it. 

```suggestion
			// If the order status is neither pending, failed, nor cancelled, the order must have gone through.
			$did_order_succeed = ! $order->has_status( array( 'failed', 'pending', 'cancelled' ) );

			/**
			 * Determine whether the cart should be cleared after payment.
			 *
			 * @since 8.7.1
			 * @param bool $should_clear_cart_after_payment Whether the cart should be cleared after payment.
			 * @param WC_Order $order The order.
			 * @return bool Whether the cart should be cleared after payment.
			 */
			$should_clear_cart_after_payment = apply_filters( 'woocommerce_should_clear_cart_after_payment', $did_order_succeed, $order );

			if ( $should_clear_cart_after_payment ) {
```

**Review by lsinger — CHANGES_REQUESTED** [review]
Thank you @alexedwardjones, that's very close to what we'd like to see happen here. 

As is, however, there's a double negation -- `$is_order_unpaid` contains a negation in its wording ("unpaid" -> "not paid") and then later when passing it into `apply_filter` we negate that `!` again. 

This makes it harder to reason about what is going on, so I'd like us to remove the double negation. 

I left a suggestion as to how to go about this, but I'm open to other ideas.

File: plugins/woocommerce/includes/wc-cart-functions.php
Comment (alexedwardjones) [reply plugins/woocommerce/includes/wc-cart-functions.php]: Thanks, I've gone ahead and applied your suggestion.

File: plugins/woocommerce/changelog/add filter for pre payment order statuses
Comment (lsinger) [inline plugins/woocommerce/changelog/add filter for pre payment order statuses]:
```suggestion
Add the `woocommerce_should_clear_cart_after_payment` filter to influence whether the cart should be cleared after payment.
```

**Comment by lsinger** [comment]
@alexedwardjones could you update the testing instructions? I cannot currently reproduce the before and after steps. I'm adding the filter like so `add_filter( 'woocommerce_should_clear_cart_after_payment', '__return_false', 10, 2 );` but am not sure what payment method(s) would actually lead to that code path being executed. For things like direct bank transfer etc. there is never a `WC()->session->order_awaiting_payment`.

**Comment by lsinger** [comment]
FWIW, I was thinking something like this would be good to have so that in testing we can again validate things work as expected: 

1. install [the Code Snippets plugin](https://wordpress.org/plugins/code-snippets/)
2. install PAYMENT_METHOD // which one? what are the steps to get into a proper testing mode? 
3. add a product to your cart and go through checkout with PAYMENT_METHOD -- the cart should be empty after checkout
4. add and activate a code snippet with the following code: `add_filter( 'woocommerce_should_clear_cart_after_payment', '__return_false' );`
5. again, add a product to your cart and go through checkout with PAYMENT_METHOD -- the cart should now still contain the product even though you've completed checkout

File: plugins/woocommerce/changelog/add-filter-woocommerce-should-clear-cart-after-payment
Comment (lsinger) [inline plugins/woocommerce/changelog/add-filter-woocommerce-should-clear-cart-after-payment]:
```suggestion

Add the `woocommerce_should_clear_cart_after_payment` filter to influence whether the cart should be cleared after payment.
```

**Comment by alexedwardjones** [comment]
@lsinger 

You can achieve the necessary state by starting a payment with a method that redirects to an external payment gateway but then cancelling before finishing the payment. The simplest way is probably installing and activating the Stripe plugin, then activating AliPay via Stripe. If you try to pay with AliPay then the redirect will fail but the order will be created in the new default status.

**Comment by lsinger** [comment]
@alexedwardjones sorry, could you actually go ahead and rewrite the PR description to reflect the new solution? We'll need that in later testing stages even after merging the PR.

**Comment by alexedwardjones** [comment]
@lsinger Hey, I've updated the test instructions and gone in a slightly different direction to instead test that the code change works by forcing the cart to empty. This avoids the need to have someone install and configure an external payment gateway.

**Comment by lsinger** [comment]
> @lsinger Hey, I've updated the test instructions and gone in a slightly different direction to instead test that the code change works by forcing the cart to empty. This avoids the need to have someone install and configure an external payment gateway.

Thanks @alexedwardjones -- that's more along the lines of what we need. I do find it rather removed from the original intent now, though. Isn't the idea that ...

1) an extension might add a custom order status and 
2) the extension might have a need to 
  a) keep that order status from clearing the cart or
  b) force that order status to clear the cart? 

While the above instructions might work, I can't really see the intent of the change in it anymore. I'd prefer if that were still visible so that when many years down the line someone else comes across this PR and wonders _why_ they will be able to understand. 🤔 Do you have a suggestion on how we could rewrite those instructions to ensure the business intent is captured?

**Comment by lsinger** [comment]
(If communication via this PR is too slow and you'd like to speed things up, please reach out to me in [the WooCommerce Community Slack](https://woocommerce.com/community-slack/).)

**Comment by alexedwardjones** [comment]
Hey @lsinger, I've updated the test instructions again. I'll message you on Slack and we can continue this discussion there.

**Comment by alexedwardjones** [comment]
Hey @vedanshujain 👋 Any chance I can get a review on this?

**Comment by alexedwardjones** [comment]
Hey @lsinger @barryhughes, I don't seem to be getting any response from @vedanshujain. Can I get another reviewer assigned to this PR.

File: plugins/woocommerce/includes/wc-cart-functions.php
Comment (barryhughes) [inline plugins/woocommerce/includes/wc-cart-functions.php]: I apologize if I missed earlier discussion about this, but might it be confusing (for other developers) that they can exercise control over cart clearance *here,* but not if execution enters the preceding block (I'm referring to the other `WC->cart->empty_cart()` call a few lines further up)?

Comment (vedanshujain) [reply plugins/woocommerce/includes/wc-cart-functions.php]: If we want to make this filter more specific, perhaps we can rename the filter to `woocommerce_clear_cart_after_payment_status_list`.

File: plugins/woocommerce/includes/wc-cart-functions.php
Comment (vedanshujain) [inline plugins/woocommerce/includes/wc-cart-functions.php]:
```suggestion
			 * @since 9.2.0
```

File: plugins/woocommerce/includes/wc-cart-functions.php
Comment (alexedwardjones) [reply plugins/woocommerce/includes/wc-cart-functions.php]: I'm happy to change the filter name if that works for folks.

Comment (vedanshujain) [reply plugins/woocommerce/includes/wc-cart-functions.php]: We discussed this offline a bit as well, Barry's suggestion makes sense here that we shouldn't have two different hooks, or a hook that fires some of the time. Can we rework this function a bit and clear cart only once at the end, behind a filter?

Comment (alexedwardjones) [reply plugins/woocommerce/includes/wc-cart-functions.php]: Thanks for the feedback 🙇  I've tried reworking the function in line with what I think you're suggesting.

Take a look and let me know if it's any closer to what you're looking for.

**Comment by barryhughes** [comment]
Feel free to request a fresh review once ready (I noticed the recent commits, but wasn't sure if you were perhaps still iterating on this).

File: plugins/woocommerce/includes/wc-cart-functions.php
Comment (barryhughes) [inline plugins/woocommerce/includes/wc-cart-functions.php]: This function is called on almost every (frontend) request, because it's hooked into `template_redirect` and, for most of those requests, `$order` will not be defined (since it's only defined under certain conditions, within the preceding if blocks).

1. This will lead to a lot of error log noise *('undefined variable $order').*
2. It might be problematic for extensions hooking into this, if they expect two parameters with the second being a well-formed `WC_Order` instance.

File: plugins/woocommerce/includes/wc-cart-functions.php
Comment (barryhughes) [inline plugins/woocommerce/includes/wc-cart-functions.php]:
```suggestion
	 * @since 9.3.0
```

**Review by barryhughes — CHANGES_REQUESTED** [review]
We should also merge the latest changes from trunk, or rebase, and solve the noted conflict.

File: plugins/woocommerce/includes/wc-cart-functions.php
Comment (barryhughes) [inline plugins/woocommerce/includes/wc-cart-functions.php]:
```suggestion
```

No need for a `@return` tag when documenting hooks [[*]](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#4-hooks-actions-and-filters).

File: plugins/woocommerce/includes/wc-cart-functions.php
Comment (alexedwardjones) [reply plugins/woocommerce/includes/wc-cart-functions.php]: Good catch.

File: plugins/woocommerce/includes/wc-cart-functions.php
Comment (barryhughes) [inline plugins/woocommerce/includes/wc-cart-functions.php:173]: **(1/4)** Thanks for your patience while we worked through this. 

I'd like to propose a final amendment, after which we can probably go ahead and merge, but let me know if this would be problematic in any way (cc @vedanshujain, I'd appreciate your eyes on this, as I may be being over-cautious).

In outline, here's what I'm thinking. Before this change, the function worked something like this:

- Run on every `template_redirect` ⚠️ 
- If `order-received` and we have a valid order instance, clear the cart.
- If we have an order awaiting payment, and the status is no longer `pending`, `failed` or `cancelled`, also clear the cart.

That was pretty efficient ... but now we have the following: 

- Run on every `template_redirect` ⚠️ 
- If `order-received` and we have a valid order instance, consider clearing the cart.
- Or, if we have an order awaiting payment, and the status is no longer `pending`, `failed` or `cancelled`, also consider clearing the cart.
- Take our decision, let third parties filter it, then actually clear the cart or not.

It's definitely not awful, but because this is running on every single `template_redirect` (hence my little ⚠️ icons), I'm wondering if we can introduce some extra efficiency gains, without too much work:

```suggestion
	$should_clear_cart_after_payment = false;
	$after_payment                   = false;
```

This new var tracks if we think a payment just took place. See next comment.

File: plugins/woocommerce/includes/wc-cart-functions.php
Comment (barryhughes) [inline plugins/woocommerce/includes/wc-cart-functions.php:186]: **(2/4)** Then here, we of course believe a payment just took place so we update the new var.

```suggestion
				$after_payment                   = true;
				$should_clear_cart_after_payment = true;
```

File: plugins/woocommerce/includes/wc-cart-functions.php
Comment (barryhughes) [inline plugins/woocommerce/includes/wc-cart-functions.php]: **(3/4)** Similarly here, we believe a payment took place. Let's make the same update (I'm also collapsing the two existing lines into one, but we don't have to do that).

```suggestion
			$should_clear_cart_after_payment = ! $order->has_status( array( 'failed', 'pending', 'cancelled' ) );
			$after_payment                   = true;
```

File: plugins/woocommerce/includes/wc-cart-functions.php
Comment (barryhughes) [inline plugins/woocommerce/includes/wc-cart-functions.php:202]: **(4/4)** Then, here, we bail out early if we don't believe a payment took place:

```suggestion
	// If it doesn't look like a payment happened, bail early.
	if ( ! $after_payment ) {
		return;
	}
```

This means we don't invoke `woocommerce_should_clear_cart_after_payment` on every `template_redirect`, but still invoke it in a way that meets the stated use case. It's perhaps verging on micro-optimization territory, but it may also offer some extra safety against edge cases where `wc()->cart` is not initialized, etc.

**Review by barryhughes — COMMENTED** [review]
One more change *to consider*, though it may not be strictly necessary.

File: plugins/woocommerce/includes/wc-cart-functions.php
Comment (alexedwardjones) [reply plugins/woocommerce/includes/wc-cart-functions.php:173]: Makes sense, I'll hold off on making the changes until @vedanshujain has taken a look.

Comment (barryhughes) [reply plugins/woocommerce/includes/wc-cart-functions.php:173]: I think we can proceed with this adjustment without waiting (Vedanshu probably won't be able to look at this soon), and we will have some space to correct if necessary as we're at a relatively early stage within the 9.3 cycle.

Comment (alexedwardjones) [reply plugins/woocommerce/includes/wc-cart-functions.php:173]: I've gone ahead and made those changes. Let me know what you think.

**Review by barryhughes — APPROVED** [review]
Great, thank you for working on this.

- Testing instructions are comprehensive, thank you.
- Per the last update, verified that the new hook is not triggered unless there's a reason to believe payment just took place.
- When the above is true, the hook is indeed fired.

**Comment by barryhughes** [comment]
Running CI, and if all passes we can go ahead and merge (this should then ship as part of the upcoming 9.3 release, scheduled for 2024-09-10).

**Comment by lsinger** [comment]
Hooray! Congrats, @alexedwardjones! 🎉

**Comment by alexedwardjones** [comment]
We made it! Thanks for your help @lsinger @barryhughes @vedanshujain

**Comment by rodelgc** [comment]
Removing `impact: high` label after clarifying certain aspects with PR reviewers.

**Comment by barryhughes** [comment]
## Alternative testing steps

Since testing with AliPay may be problematic in some cases, here are some alternative steps.

#### 1) Confirm normal behaviour

This PR should not change any of the expected, default checkout flows. So, run through some test purchases: they should work exactly as you expect/exactly as before this change was introduced.

#### 2) Create a test gateway that does not automatically clear the cart

Lots of gateways, including default ones such as BACS or Cheque Payments, take care of clearing the cart after the order is placed. So, they are not suitable for testing this change. Let's instead create a custom payment gateway of our own.

Create a new file in a suitable location. I recommend `wp-content/mu-plugins/test-44515.php` (remove it when you finish testing this PR), containing the following:

```php
<?php

add_action( 'woocommerce_init', function () {
	class Test_PR_44515_Gateway extends WC_Payment_Gateway {
		public function __construct() {
			$this->id                 = 'test-pr-44515';
			$this->method_title       = 'Test PR 44515';
			$this->method_description = 'A payment gateway that can be used to test PR#44515.';
			$this->has_fields         = false;
			$this->init_form_fields();
			$this->init_settings();
			$this->title       = $this->get_option( 'title' );
			$this->description = $this->get_option( 'description' );
			add_action( 'woocommerce_update_options_payment_gateways_' . $this->id, array(
				$this,
				'process_admin_options'
			) );
		}

		public function init_form_fields() {
			$this->form_fields = array(
				'enabled'     => array(
					'title'   => 'Enable/Disable',
					'type'    => 'checkbox',
					'label'   => 'Enable Test Gateway',
					'default' => 'yes'
				),
				'title'       => array(
					'title'       => 'Title',
					'type'        => 'text',
					'description' => 'This controls the title which the user sees during checkout.',
					'default'     => 'Test PR #44515',
					'desc_tip'    => true,
				),
				'description' => array(
					'title'       => 'Description',
					'type'        => 'textarea',
					'description' => 'This controls the description which the user sees during checkout.',
					'default'     => 'This is our test gateway..',
				),
			);
		}

		public function process_payment( $order_id ) {
			$order = wc_get_order( $order_id );
			$order->update_status( 'complete' );

			return array(
				'result'   => 'success',
				'redirect' => $this->get_return_url( $order ),
			);
		}
	}

	add_filter('woocommerce_payment_gateways', function ( $gateways ) {
		$gateways[] = Test_PR_44515_Gateway::class;
		return $gateways;
	});
} );
```

Now visit **WooCommerce ‣ Settings ‣ Payments** and enable the **Test PR 44515** gateway. Note that this is a very basic gateway for testing purposes only and, to see it on the frontend, you will need to use the *classic checkout*.

As a reminder, this is simply a page containing the `[woocommerce_checkout]` shortcode (you'll need to use a *shortcode block* to add this from the block editor. Create that if you need to, and copy the page URL.

Now, acting as a shopper, add something to your cart then place the order *via the classic checkout*. If multiple payment gateways are enabled, be sure to select our new test gateway! Everything should work as expected and if you subsequently visit the cart page, you should see it is empty (again, as expected).

#### 3) Verify the new filter hook works

Amend your test code, and add the following line to the bottom (on a new line, below the final `} );`):

```php
add_filter( 'woocommerce_should_clear_cart_after_payment', fn () => false );
```

Now repeat that last test, again taking care to place the order via the classic checkout page. This time, if you visit the cart page after placing the order, you should find the cart is still populated with whatever item(s) you just purchased.

**Pull request merged**
